File: IPWorks EDI Translator Samples/X12 Parser/python/x12_parser.py
# 
# IPWorks EDI Translator 2022 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks EDI Translator in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworkseditranslator
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworkseditranslator import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] == None:
    args[index] = input(prompt)

import sys
import string
from ipworkseditranslator import *

input = sys.hexversion<0x03000000 and raw_input or input

## Define events!
def endFunctionalGroup(e):
  print("EndFunctionalGroup: " + e.tag)

def endInterchange(e):
  print("EndInterchange: " + e.tag)

def endLoop(e):
  print("EndLoop")

def endTransaction(e):
  print("EndTransaction: " + e.tag)

def error(e):
  print("ERROR: " + e.error_code + ':' + e.description)

def resolveSchema(e):
  print("ResolveSchema: " + e.transaction_code)

def segment(e):
  print("Segment: " + e.name)

def startFunctionalGroup(e):
  print("StartFunctionalGroup: " + e.tag)

def startInterchange(e):
  print("StartInterchange: " + e.tag)

def startLoop(e):
  print("StartLoop: " + e.name)

def startTransaction(e):
  print("StartTransaction: " + e.tag)

def warning(e):
  print("WARNING: " + e.warn_code + ": " + e.message)


## Check Command Line Args
if len(sys.argv) != 2:
  print("usage: x12parser file")
  print("")
  print("  file the x12 file to parse")
  print("\r\nExample: py x12parser.py ./x12.txt")
  sys.exit()

## Set up reader
x12reader1 = X12Reader();

## Add try to catch errors
try:
  ## Add events to reader
  x12reader1.on_end_functional_group = endFunctionalGroup
  x12reader1.on_end_interchange = endInterchange
  x12reader1.on_end_loop = endLoop
  x12reader1.on_end_transaction = endTransaction
  x12reader1.on_error = error
  x12reader1.on_resolve_schema = resolveSchema
  x12reader1.on_segment = segment
  x12reader1.on_start_functional_group = startFunctionalGroup
  x12reader1.on_start_interchange = startInterchange
  x12reader1.on_start_loop = startLoop
  x12reader1.on_start_transaction = startTransaction
  x12reader1.on_warning = warning

  ## Set up encoding and schema
  x12reader1.config("Encoding=iso-8859-1")
  x12reader1.set_schema_format(6);
  x12reader1.load_schema("./RSSBus_00401_810.json");

  ## Load in the input file and parse()
  x12reader1.set_input_file(sys.argv[1])
  x12reader1.parse()

except Exception as e:
  print(e)

File: IPWorks EDI Translator Samples/X12 Parser/netcore/x12parser.cs
/*
 * IPWorks EDI Translator 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks EDI Translator in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworkseditranslator
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using nsoftware.IPWorksEDITranslator;

class x12parserDemo
{
  private static X12Reader x12reader = new nsoftware.IPWorksEDITranslator.X12Reader();

  static void Main(string[] args)
  {
    if (args.Length < 2)
    {
      Console.WriteLine("usage: x12parser /f file");
      Console.WriteLine("  file         the x12 file to parse");
      Console.WriteLine("\nExample: x12parser /f ./x12.txt");
    }
    else
    {
      x12reader.OnEndFunctionalGroup += (s, e) => { Console.WriteLine("EndFunctionalGroup: " + e.Tag); };
      x12reader.OnEndInterchange += (s, e) => { Console.WriteLine("EndInterchange: " + e.Tag); };
      x12reader.OnEndLoop += (s, e) => { Console.WriteLine("EndLoop"); };
      x12reader.OnEndTransaction += (s, e) => { Console.WriteLine("EndTransaction: " + e.Tag); };
      x12reader.OnError += (s, e) => { Console.WriteLine("Error " + e.ErrorCode + ": " + e.Description); };
      x12reader.OnResolveSchema += (s, e) => { Console.WriteLine("ResolveSchema: " + e.TransactionCode); };
      x12reader.OnSegment += (s, e) => { Console.WriteLine("Segment: " + e.Name); };
      x12reader.OnStartFunctionalGroup += (s, e) => { Console.WriteLine("StartFunctionalGroup: " + e.Tag); };
      x12reader.OnStartInterchange += (s, e) => { Console.WriteLine("StartInterchange: " + e.Tag); };
      x12reader.OnStartLoop += (s, e) => { Console.WriteLine("StartLoop: " + e.Name); };
      x12reader.OnStartTransaction += (s, e) => { Console.WriteLine("StartTransaction: " + e.Tag); };
      x12reader.OnWarning += (s, e) => { Console.WriteLine("Warning " + e.WarnCode + ": " + e.Message); };

      try
      {
        System.Collections.Generic.Dictionary<string, string> myArgs = ConsoleDemo.ParseArgs(args);

        x12reader.InputFile = myArgs["f"];

        x12reader.Config("Encoding=iso-8859-1");
        x12reader.SchemaFormat = X12ReaderSchemaFormats.schemaJSON;
        x12reader.LoadSchema("../../../RSSBus_00401_810.json");

        // This demo provides information about the parsed document through the events.
        // To navigate the document using the XPath property, first set:
        // x12reader.Config("ResolveXPathOnSet=true");
        x12reader.Parse();
      }
      catch (Exception ex)
      {
        Console.WriteLine(ex.Message);
      }
    }
  }
}




class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add an key to the dictionary for each argument
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/" then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}